### ISSUE TITLE
Marginals don't handle infinite numerical values in support

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#179]
At present:

````Enumerate(function() { flip(0.5) ? -Infinity : Infinity }).support(); === [ -Infinity ]````

Rather than `[ -Infinity, Infinity]`.

This happens because `JSON.stringify(Infinity) === JSON.stringify(-Infinity) === 'null'`.

See discussion [here](https://github.com/probmods/webppl/pull/162#issuecomment-138392236).

**Comment by stuhlmueller** [comment]
This is only one of a number of issues arising from the use of [JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Description):

For example, this returns an ERP with probability 1 on `{}`:

    var f = function(){
      if (flip(.5)){
        return { x: undefined, y: function(){} };
      } else {
        return {};
      }
    }
    Enumerate(f)

**Comment by null-a** [comment]
Also see #108.

**Comment by stuhlmueller** [comment]
Maybe we should use [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)?

**Comment by null-a** [comment]
> Maybe we should use Map?

I was thinking that.

One potential snag is the way it [compares keys](http://www.ecma-international.org/ecma-262/6.0/#sec-samevaluezero) - I suspect this means it doesn't help with #108. (And I don't think this is something which can be customized yet.) If we care about #108 perhaps we could extend `Map` though?

**Comment by null-a** [comment]
> I suspect this means it doesn't help with #108 

Only identical objects are considered equal, so I think it's actually problematic for stuff which currently works.

**Comment by stuhlmueller** [comment]
> Only identical objects are considered equal, so I think it's actually problematic for stuff which currently works.

Right. I think what we want is a hash table that uses [`_.isEqual`](http://underscorejs.org/#isEqual) for equality, and something reasonably efficient for hashing.
